File: 1577.number-of-ways-where-square-of-number-is-equal-to-product-of-two-numbers.py
#
# @lc app=leetcode id=1577 lang=python3
#
# [1577] Number of Ways Where Square of Number Is Equal to Product of Two Numbers
#

# @lc code=start
from typing import List
class Solution:
    def numTriplets(self, nums1: List[int], nums2: List[int]) -> int:
        #No Unique requirement
        #i

        #two pointer helper function



        def count_triplet(n1: List[int], n2: List[int]) -> int:
            count = 0
             #! the first approach works, but TLE for many duplicated pairs
            # for i in range(len(n1)):
            #     target = n1[i] ** 2
            #     for j in range(len(n2)):
            #         right =  len(n2) - 1
            #         while j < right:
            #             cur = n2[j] * n2[right]
            #             if cur == target:
            #                 count += 1
            #                 right -= 1
            #             elif cur > target:
            #                 right -= 1
            #             else:
            #                 break
            # return count

            #! Optimized approach: jumping on the duplicated pairs
            for i in range(len(n1)):
                cnt = 0
                target = n1[i] ** 2

                left = 0
                right = len(n2) - 1

                while left < right:
                    cur = n2[left] * n2[right]
                    #Below is all mathematical counting logic
                    if cur == target:
                        if n2[left] == n2[right]:
                            count += (right - left + 1) * (right - left) // 2
                            break
                        else:
                            left_cnt = 1
                            right_cnt = 1
                            while left + 1 < right and n2[left] == n2[left + 1]:
                                left += 1
                                left_cnt += 1
                            while right - 1 > left and n2[right] == n2[right - 1]:
                                right -= 1
                                right_cnt += 1
                            count += left_cnt * right_cnt
                            left += 1
                            right -= 1
                    elif cur > target:
                        right -= 1
                    else:
                        left += 1
            return count
        nums1.sort()
        nums2.sort()
        return count_triplet(nums1, nums2) + count_triplet(nums2, nums1)



        
# @lc code=end

File: 1577.number-of-ways-where-square-of-number-is-equal-to-product-of-two-numbers.java
/*
 * @lc app=leetcode id=1577 lang=java
 *
 * [1577] Number of Ways Where Square of Number Is Equal to Product of Two Numbers
 */
import java.util.*;
// @lc code=start
class Solution {
    public int numTriplets(int[] nums1, int[] nums2) {
        // return getCnt(nums1, nums2) + getCnt(nums2, nums1);
        Arrays.sort(nums1);
        Arrays.sort(nums2);
        return getCnt(nums1, nums2) + getCnt(nums2, nums1);
        
    }
    // private int getCnt(int[] nums1, int[] nums2) {
    //     Map<Long, Integer> map = new HashMap<>();
    //     for (int n : nums2) {
    //         long square = (long)n * n;
    //         map.put(square, map.getOrDefault(square, 0) + 1);
    //     }
    //     int cnt = 0;
    //     for (int i = 0; i < nums1.length; i++) {
    //         for (int j = i + 1; j < nums1.length; j++) {
    //             long product = (long)nums1[i] * nums1[j];
    //             cnt += map.getOrDefault(product, 0);
    //         }
    //     }
    //     return cnt;
    // }
    private int getCnt(int[] nums1, int[] nums2){
        int cnt = 0;
        for(int i = 0; i < nums1.length;i++){
            long square = (long) nums1[i] * nums1[i];
            int left = 0, right = nums2.length -1;
            while(left < right){
                long product = (long) nums2[left] * nums2[right];
                if(product > square){
                    right--;
                } else if(product < square){
                    left++;
                } else {
                    if(nums2[left] == nums2[right]){
                        int size = right - left + 1;
                        cnt += size * (size - 1) / 2;
                        break;
                    }
                    else{
                        int leftCnt = 1, rightCnt = 1;
                        while(left + 1 < right && nums2[left] == nums2[left + 1]){
                            leftCnt++;
                            left++;
                        }
                        while(right - 1 > left && nums2[right] == nums2[right - 1]){
                            rightCnt++;
                            right--;
                        }
                        cnt += leftCnt * rightCnt;
                        left++;
                        right--;
                    }
                }
            }
        }
        return cnt;
    }
    
}
// @lc code=end

